### DISCUSSION TITLE
Proper evaluation for MILP

### DISCUSSION THREAD

**Discussion by drdv** [optimatika/ojAlgo#417] [General, ojAlgo and related code]
I wonder what is the bottleneck when solving MILPs. I didn't find an evaluation against free alternatives. For LPs there is a comparison but with Gurobi (and only for small problems due to the limitations of the free version).

**Comment by apete** [comment]
Presumably you can find inefficiencies "everywhere". A key thing completely missing is cut-generation. In more general terms ojAlgo needs to be better at prioritising which nodes to evaluate (first) and ultimately if it needs to be evaluated at all.

To get to know how the `IntegerSolver` works I would:

- Find a small small problem to investigate
- Turn on debug logging `model.options.debug(IntegerSolver.class)`
- Run in a debugger with breakpoints in the `IntegerSolver.compute(...)` method

**Reply by apete** [reply]
There is now some preliminary work to improve the IntegerSolver: https://github.com/optimatika/ojAlgo/tree/IntegerStrategy

Primarily this changes how the solver is multi-threaded, and also makes it much more configurable.

Still no cuts...

**Reply by drdv** [reply]
Thanks, I still haven't had time to look into the `IntegerSolver`.
